miio/roidmivacuum_miot.py

```diff
@@ -0,0 +1,769 @@
+            {'did': 'double_clean', 'siid': 8, 'piid': 20},
+            {'did': 'edge_sweep', 'siid': 8, 'piid': 21},
+            {'did': 'led_switch', 'siid': 8, 'piid': 22}
+            {'did': 'lidar_collision', 'siid': 8, 'piid': 23},
+            {'did': 'station_key', 'siid': 8, 'piid': 24},
+            {'did': 'station_led', 'siid': 8, 'piid': 25},
+            {'did': 'current_audio', 'siid': 8, 'piid': 26},
+            {'did': 'progress', 'siid': 8, 'piid': 28},
+            {'did': 'station_type', 'siid': 8, 'piid': 29},
+            {'did': 'volume', 'siid': 9, 'piid': 1},
+            {'did': 'mute', 'siid': 9, 'piid': 2}
+        ]
+
+        """
+        self.data = data
+
+    @property
+    def battery(self) -> int:
+        """Remaining battery in percentage."""
+        return self.data["battery_level"]
+
+    @property
+    def error_code(self) -> int:
+        """Error code as returned by the device."""
+        return int(self.data["error_code"])
+
+    @property
+    def error(self) -> str:
+        """Human readable error description, see also :func:`error_code`."""
+        try:
+            return error_codes[self.error_code]
+        except KeyError:
+            return "Definition missing for error %s" % self.error_code
+
+    @property
+    def charging_state(self) -> ChargingState:
+        """Charging state (Charging/Discharging)"""
+        try:
+            return ChargingState(self.data["charging_state"])
+        except ValueError:
+            _LOGGER.error("Unknown ChargingStats (%s)", self.data["charging_state"])
+            return ChargingState.Unknown
+
+    @property
+    def sweep_mode(self) -> SweepMode:
+        """Sweep mode point/area/total etc."""
+        try:
+            return SweepMode(self.data["sweep_mode"])
+        except ValueError:
+            _LOGGER.error("Unknown SweepMode (%s)", self.data["sweep_mode"])
+            return SweepMode.Unknown
+
+    @property
+    def fan_speed(self) -> FanSpeed:
+        """Current fan speed."""
+        try:
+            return FanSpeed(self.data["fanspeed_mode"])
+        except ValueError:
+            _LOGGER.error("Unknown FanSpeed (%s)", self.data["fanspeed_mode"])
+            return FanSpeed.Unknown
+
+    @property
+    def sweep_type(self) -> SweepType:
+        """Current sweep type sweep/mop/sweep&mop."""
+        try:
+            return SweepType(self.data["sweep_type"])
+        except ValueError:
+            _LOGGER.error("Unknown SweepType (%s)", self.data["sweep_type"])
+            return SweepType.Unknown
+
+    @property
+    def path_mode(self) -> PathMode:
+        """Current path-mode:  normal/y-mopping etc."""
+        try:
+            return PathMode(self.data["path_mode"])
+        except ValueError:
+            _LOGGER.error("Unknown PathMode (%s)", self.data["path_mode"])
+            return PathMode.Unknown
+
+    @property
+    def mop_present(self) -> bool:
```

Comment:
```suggestion
    def is_water_box_attached(self) -> bool:
```
is used by rockrobo vacuums for exactly the same thing, right?

Reply:
On Roidme Eve the water-box is integrated in the dust-bin. The mop-module is approximately 1 cm high and can be attached below the combined dust+water-box.
Strictly speaking the mop is only a piece of fabric, and Roidmi Eve does not have a sensor to detect if that is pressent (and I guess that this is the same for Roborock).
If you would like a common name for Roborock and Roidme Eve, then we might be able to argue that the water-box on Roborock is a mop-module?
For now I think that best name for Roidme Eve it "is_mop_module_attached"?

Reply:
Ah, so this is really about the mop, so I'd just use `is_mop_attached`.